onnxruntime/contrib_ops/cuda/quantization/attention_quantization.cc

```diff
@@ -0,0 +1,189 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "attention_quantization.h"
+#include "attention_quantization_impl.cuh"
+#include "contrib_ops/cuda/bert/attention_impl.h"
+#include "core/framework/tensorprotoutils.h"
+#include "core/providers/common.h"
+#include "core/providers/cuda/cuda_common.h"
+#include "core/providers/cuda/shared_inc/fpgeneric.h"
+#include "core/providers/cuda/shared_inc/integer_gemm.h"
+#include "core/providers/cuda/tensor/quantize_linear.h"
+
+using namespace onnxruntime::cuda;
+using namespace onnxruntime::common;
+
+namespace onnxruntime {
+namespace contrib {
+namespace cuda {
+
+#define REGISTER_KERNEL_TYPED(T, TQuant)                                 \
+  ONNX_OPERATOR_TYPED_KERNEL_EX(                                         \
+      QAttention,                                                        \
+      kMSDomain,                                                         \
+      1,                                                                 \
+      T##_##TQuant,                                                      \
+      kCudaExecutionProvider,                                            \
+      KernelDefBuilder()                                                 \
+          .InputMemoryType<OrtMemTypeCPUInput>(3)                        \
+          .InputMemoryType<OrtMemTypeCPUInput>(4)                        \
+          .InputMemoryType<OrtMemTypeCPUInput>(6)                        \
+          .InputMemoryType<OrtMemTypeCPUInput>(7)                        \
+          .TypeConstraint("T1", DataTypeImpl::GetTensorType<TQuant>())   \
+          .TypeConstraint("T2", DataTypeImpl::GetTensorType<TQuant>())   \
+          .TypeConstraint("T3", DataTypeImpl::GetTensorType<T>())        \
+          .TypeConstraint("T4", DataTypeImpl::GetTensorType<int32_t>()), \
+      QAttention<T, TQuant>);
+
+REGISTER_KERNEL_TYPED(float, int8_t)
+REGISTER_KERNEL_TYPED(MLFloat16, int8_t)
+
+template <typename T>
+Status QAttention<T, int8_t>::CheckInputs(const Tensor* input,
+                                          const Tensor* weights,
+                                          const Tensor* bias,
+                                          const Tensor* input_scale_tensor,
+                                          const Tensor* weight_scale_tensor,
+                                          const Tensor* mask_index,
+                                          const Tensor* i_zp_tensor,
```

Comment:
It might be better to change the naming:
 i_zp_tensor -> input_zero_point
 w_zp_tensor -> weight_zero_point